### ISSUE TITLE
Inconsistent implicit type casting when using `max_horizontal` and `mean_horizontal`

### ISSUE THREAD

**Issue by NicolasMuellerQC** [pola-rs/polars#15951]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

df = pl.DataFrame(data=dict(x=[1, 2, 3]))
print(
    df.with_columns(
        (1 - 0.5 * pl.col("x")).alias("y"),
        pl.max_horizontal(0, 1 - 0.5 * pl.col("x")).alias("z"),
        pl.max_horizontal(0.0, 1 - 0.5 * pl.col("x")).alias("w"),
        pl.max_horizontal(0, 1.0 - 0.5 * pl.col("x")).alias("r"),
    )
)

# Output:
# shape: (3, 5)
# ┌─────┬──────┬─────┬─────┬─────┐
# │ x   ┆ y    ┆ z   ┆ w   ┆ r   │
# │ --- ┆ ---  ┆ --- ┆ --- ┆ --- │
# │ i64 ┆ f64  ┆ i32 ┆ f64 ┆ f64 │
# ╞═════╪══════╪═════╪═════╪═════╡
# │ 1   ┆ 0.5  ┆ 1   ┆ 1.0 ┆ 0.5 │
# │ 2   ┆ 0.0  ┆ 0   ┆ 0.0 ┆ 0.0 │
# │ 3   ┆ -0.5 ┆ 0   ┆ 0.0 ┆ 0.0 │
# └─────┴──────┴─────┴─────┴─────┘


print(
    df.with_columns(
        (1 - 0.5 * pl.col("x")).alias("y"),
        pl.mean_horizontal(0, 1 - 0.5 * pl.col("x")).alias("z"),
        pl.mean_horizontal(0.0, 1 - 0.5 * pl.col("x")).alias("w"),
        pl.mean_horizontal(0, 1.0 - 0.5 * pl.col("x")).alias("r"),
    )
)
# Output:
# shape: (3, 5)
# ┌─────┬──────┬─────┬─────┬───────┐
# │ x   ┆ y    ┆ z   ┆ w   ┆ r     │
# │ --- ┆ ---  ┆ --- ┆ --- ┆ ---   │
# │ i64 ┆ f64  ┆ f64 ┆ f64 ┆ f64   │
# ╞═════╪══════╪═════╪═════╪═══════╡
# │ 1   ┆ 0.5  ┆ 0.5 ┆ 0.5 ┆ 0.25  │
# │ 2   ┆ 0.0  ┆ 0.0 ┆ 0.0 ┆ 0.0   │
# │ 3   ┆ -0.5 ┆ 0.0 ┆ 0.0 ┆ -0.25 │
# └─────┴──────┴─────┴─────┴───────┘

```

### Log output

```shell
No output.
```

### Issue description

The implicit type casting when using `max_horizontal`  is inconsistent:
- The expression `(1 - 0.5 * pl.col("x")).alias("y")` on the `df` has type f64 and correctly calculates e.g. `1-0.5*1 = 0.5` in the first row.
- The expression `pl.max_horizontal(0, 1 - 0.5 * pl.col("x")).alias("z")` has type i32 even though the second expression has type f64 (by point one). I think operations involving i32 and f64 should be casted to f64 if there is any implicit casting.
- The expression `pl.max_horizontal(0.0, 1 - 0.5 * pl.col("x")).alias("w")` has type f64, but it looks like the second column was casted to i32 before the `max` aggregation, even though it has type f64 (by point one).
- The expression `pl.max_horizontal(0, 1.0 - 0.5 * pl.col("x")).alias("r")`has type f64 and the output is what I would expect.

The same problem applies to `mean_horizontal`.

### Expected behavior

I would expect the columns `"z"`, `"w"`, `"r"` to be equal in both cases.  The column `"r"` is correct.

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.23
Index type:           UInt32
Platform:             macOS-14.4.1-arm64-arm-64bit
Python:               3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:34:54) [Clang 16.0.6 ]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          2.2.1
connectorx:           0.3.2
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               2024.3.1
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           3.8.4
nest_asyncio:         1.6.0
numpy:                1.26.4
openpyxl:             3.1.2
pandas:               2.2.2
pyarrow:              15.0.2
pydantic:             1.10.14
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           2.0.29
xlsx2csv:             <not installed>
xlsxwriter:           3.1.9

```

</details>

**Comment by TobiasDummschat** [comment]
This seems to be fixed in 0.20.24. Now all columns added columns have type `f64`.

**Comment by NicolasMuellerQC** [comment]
Thanks, I can confirm it's fixed in 0.20.24!
